### DISCUSSION TITLE
How to create a web chat app using Gemini API (await model.generateContent is not working)

### DISCUSSION THREAD

**Discussion by DefendEager** [community/community#81778] [Programming Help] [answered]
Developing with VSCode. I usually develop using flutter, so I don't know much about html or javascript.
I'd like to create **a web page where users can chat with Gemini**, but I can't seem to get it to work. What html or JS code or work should I add to the code shown below? Or maybe the development method is wrong in the first place? (There is so little information on the internet that I don't understand how to create this.) Please let me know.

**The problem I'm experiencing**
I am debugging with Chrome. No errors are shown in the debug console.
The problem is, when I type a question in the textbox and press the button, console.log("input: ", input); works fine, but The subsequent logs, console.log("result: ", result) ;, is not working, and of course there is no response from Gemini. Does that mean that communication with the API is not possible?
In addition, API_KEY = "APIKEY" is set correctly in the project.

**Project source code**

> `<!DOCTYPE html>
> <html>
>     <head>
>         <meta charset="UTF-8">
>         <title>This is a test.</title>
>     </head>
>   <body>
>     <h1>Chat with AI</h1>
>   <form name="form1">
>     <input type="text" name="input1" id="chat-input" value="" placeholder="your question">
>     <button id="button">send</button>
>   </form>
>   <div id="edit"></div>
> 
>     <script type="importmap">
>       {
>         "imports": {
>           "@google/generative-ai": "https://esm.run/@google/generative-ai"
>         }
>       }
>     </script>
>     <script type="module">
>       import { GoogleGenerativeAI } from "https://esm.run/@google/generative-ai";
>       const API_KEY = "APIKEY";
>       const genAI = new GoogleGenerativeAI(API_KEY);
>       async function run() {
>         let input = document.form1.input1.value;
>   // For text-only input, use the gemini-pro model
>   const model = genAI.getGenerativeModel({ model: "gemini-pro"});
>     console.log("input: ", input);
>     const result = await model.generateContent(input);
>     console.log("result: ", result);
>     const response = await result.response;
>     console.log("response: ", response);
>     const text = response.text();
>    console.log(text);
>    document.getElementById('edit').innerHTML = text; 
> }
> 
> window.onload = function() {
>   document.getElementById('button').addEventListener('click', run);
>  };
>  
>     </script>
>   </body>
> </html>`

There is no further code in the project other than in launch.json.
How can I make sure that the code following `const result = await model.generateContent(input);` works correctly? Thank you for your guidance.

**Referenced page**
[](https://ai.google.dev/tutorials/web_quickstart?hl=ja)

**Comment by ShadBalti** [comment]
> Developing with VSCode. I usually develop using flutter, so I don't know much about html or javascript. I'd like to create **a web page where users can chat with Gemini**, but I can't seem to get it to work. What html or JS code or work should I add to the code shown below? Or maybe the development method is wrong in the first place? (There is so little information on the internet that I don't understand how to create this.) Please let me know.
> 
> **The problem I'm experiencing** I am debugging with Chrome. No errors are shown in the debug console. The problem is, when I type a question in the textbox and press the button, console.log("input: ", input); works fine, but The subsequent logs, console.log("result: ", result) ;, is not working, and of course there is no response from Gemini. Does that mean that communication with the API is not possible? In addition, API_KEY = "APIKEY" is set correctly in the project.
> 
> **Project source code**
> 
> > `
> > 
> >     
> >         
> >         <title>This is a test.</title>
> >     
> >   
> >     # Chat with AI
> >   
> >     
> >     send
> >   
> >   
> > ```
> > <script type="importmap">
> >   {
> >     "imports": {
> >       "@google/generative-ai": "https://esm.run/@google/generative-ai"
> >     }
> >   }
> > </script>
> > <script type="module">
> >   import { GoogleGenerativeAI } from "https://esm.run/@google/generative-ai";
> >   const API_KEY = "APIKEY";
> >   const genAI = new GoogleGenerativeAI(API_KEY);
> >   async function run() {
> >     let input = document.form1.input1.value;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > // For text-only input, use the gemini-pro model
> > const model = genAI.getGenerativeModel({ model: "gemini-pro"});
> > console.log("input: ", input);
> > const result = await model.generateContent(input);
> > console.log("result: ", result);
> > const response = await result.response;
> > console.log("response: ", response);
> > const text = response.text();
> > console.log(text);
> > document.getElementById('edit').innerHTML = text;
> > }
> > window.onload = function() {
> > document.getElementById('button').addEventListener('click', run);
> > };
> > ```
> > </script>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> >   
> > `
> 
> There is no further code in the project other than in launch.json. How can I make sure that the code following `const result = await model.generateContent(input);` works correctly? Thank you for your guidance.
> 
> **Referenced page**

Hey @DefendEager 

Thanks for reaching out and sharing your experience with building a web chat using Gemini and VSCode. I understand your frustration with the lack of information online and debugging the issue you're facing. Don't worry, I'm here to help!

**Analyzing the problem:**

Based on your description, it seems like the communication with the Gemini API is breaking down after the `model.generateContent(input)` call. While console logs for input and the generation call itself work, subsequent logs for response and text remain silent. This suggests an issue with receiving the generated content from the API.

**Here are some possibilities:**

1. **Incorrect response object access:** You're currently using `await result.response;` and then `response.text();`. Double-check the exact structure of the returned object from `result`. The actual text might be nested within a different property instead of directly under `response`. Try exploring the object properties using console logs to identify where the text resides.

2. **Network error:** While API_KEY is set correctly, ensure network communication is active and successful. Check Chrome devtools for any network errors related to the Gemini API call.

3. **Model initialization:** The provided code snippet initializes the model within the `run()` function. This might cause re-initialization each time the button is clicked, potentially leading to unexpected behavior. Try moving the model initialization outside the function (before `async function run()`) to ensure a single instance persists throughout the interaction.

4. **Error handling:** The current code doesn't seem to explicitly handle potential errors during the `await` calls. Add try-catch blocks around `await model.generateContent(input);` and subsequent `await` calls to capture any errors and log them for debugging.

**Additional tips:**

* Consider using asynchronous callbacks instead of direct console logs for tracking the response progression. This can offer a cleaner and more controlled flow.
* Refer to the official Gemini API documentation for detailed information on the response object structure and expected behavior: [https://cloud.google.com/natural-language/docs](https://cloud.google.com/natural-language/docs)



**I hope this helpful!**

**Reply by itseffi** [reply]
TOP!  :)

**Reply by DefendEager** [reply, marked as answer]
I solved this problem by changing `run(){…}` into `run(e){e.preventDefault(); …}`, and that’s all! But **_I don’t know why it fixed the issue. Is there anyone who knows?_**
However, I have eventually chosen to use Node.js and Express.js, considering the security of my API key, so I don’t need the solution any more😭.
Unless you deploy onto the Internet, my code above seems to be the easiest way to play around with Gemini in your localhost (web) environment👍🏻.

**Comment by samus-aran** [comment]
Hi @DefendEager , Thanks for joining us for the GitHub Universe celebrations in the GitHub Community, we're glad you're here! 

We want to make sure you’re getting the best support you can now that GitHub Universe has ended so I've gone ahead and moved this to the most relevant category for you.

Thank you for helping us maintain a productive and tidy community for all our members.
